modules/core/src/main/java/org/apache/synapse/mediators/transform/pfutils/FreeMarkerTemplateProcessor.java

```diff
@@ -0,0 +1,415 @@
+
+        usingPayload = templateString.contains(PAYLOAD_INJECTING_NAME);
+        usingArgs = templateString.contains(ARGS_INJECTING_NAME);
+        usingPropertyCtx = templateString.contains(CTX_PROPERTY_INJECTING_NAME);
+        usingPropertyAxis2 = templateString.contains(AXIS2_PROPERTY_INJECTING_NAME);
+        usingPropertyTransport = templateString.contains(TRANSPORT_PROPERTY_INJECTING_NAME);
+    }
+
+    /**
+     * Inject argument values to freemarker
+     *
+     * @param messageContext Message context
+     * @param mediaType      Output media type
+     * @param data           Freemarker data input
+     */
+    private void injectArgs(MessageContext messageContext, String mediaType, Map<String, Object> data) {
+
+        if (usingArgs) {
+            Map<String, Object> argsValues = new HashMap<>();
+            Map<String, ArgumentDetails>[] argValues = getArgValues(mediaType, messageContext);
+            for (int i = 0; i < argValues.length; i++) {
+                Map<String, ArgumentDetails> argValue = argValues[i];
+                Map.Entry<String, ArgumentDetails> argumentDetailsEntry = argValue.entrySet().iterator().next();
+                String replacementValue = prepareReplacementValue(mediaType, messageContext, argumentDetailsEntry);
+                argsValues.put(ARGS_INJECTING_PREFIX + (i + 1), replacementValue);
+            }
+            data.put(ARGS_INJECTING_NAME, argsValues);
+        }
+    }
+
+    /**
+     * Inject  payload in to FreeMarker
+     *
+     * @param messageContext MessageContext
+     * @param payloadType    Input payload type
+     * @param data           FreeMarker data input
+     * @throws SAXException
+     * @throws IOException
+     * @throws ParserConfigurationException
+     */
+    private void injectPayloadVariables(MessageContext messageContext, int payloadType, Map<String, Object> data)
+            throws SAXException, IOException, ParserConfigurationException {
+
+        if (usingPayload) {
+            if (payloadType == XML_PAYLOAD_TYPE) {
+                injectXmlPayload(messageContext, data);
+            } else if (payloadType == JSON_PAYLOAD_TYPE) {
+                injectJsonPayload((Axis2MessageContext) messageContext, data);
+            } else if (payloadType == TEXT_PAYLOAD_TYPE) {
+                injectTextPayload(messageContext, data);
+            } else {
+                data.put(PAYLOAD_INJECTING_NAME, "");
+            }
+        }
+    }
+
+    /**
+     * Inject a JSON payload in the FreeMarker
+     *
+     * @param messageContext Message context
+     * @param data           FreeMarker data input
+     */
+    private void injectJsonPayload(Axis2MessageContext messageContext, Map<String, Object> data) {
+
+        org.apache.axis2.context.MessageContext axis2MessageContext = messageContext.getAxis2MessageContext();
+        String jsonPayloadString = JsonUtil.jsonPayloadToString(axis2MessageContext);
```

Comment:
Can this be a JSONPrimitive ex:- following are valid JSON payloads 
boolean - true or false
single number
single string
null

Reply:
fixed this using gson and checking for JSON primitives.